apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import glob from 'glob';
+import { performance } from 'perf_hooks';
+import { AlreadyExistsBehavior, FileSystem, LegacyAdapters } from '@rushstack/node-core-library';
+import { TapOptions } from 'tapable';
+
+import { IHeftPlugin } from '../pluginFramework/IHeftPlugin';
+import { HeftSession } from '../pluginFramework/HeftSession';
+import { HeftConfiguration } from '../configuration/HeftConfiguration';
+import { ScopedLogger } from '../pluginFramework/logging/ScopedLogger';
+import { Async } from '../utilities/Async';
+import {
+  IHeftEventActions,
+  CoreConfigFiles,
+  HeftEvent,
+  IExtendedSharedCopyConfiguration
+} from '../utilities/CoreConfigFiles';
+import {
+  IBuildStageContext,
+  IBundleSubstage,
+  ICompileSubstage,
+  IPostBuildSubstage,
+  IPreCompileSubstage
+} from '../stages/BuildStage';
+
+const globEscape: (unescaped: string[]) => string[] = require('glob-escape'); // No @types/glob-escape package exists
+
+const PLUGIN_NAME: string = 'CopyFilesPlugin';
+const HEFT_STAGE_TAP: TapOptions<'promise'> = {
+  name: PLUGIN_NAME,
+  stage: Number.MAX_SAFE_INTEGER / 2 // This should give us some certainty that this will run after other plugins
+};
+
+const MAX_PARALLELISM: number = 100;
+
+export interface ICopyFileDescriptor {
+  sourceFilePath: string;
+  destinationFilePath: string;
```

Comment:
It'd be more efficient for copy (i.e. - not hardlink) operations to read the file once and write it `n` times instead of repeatedly calling `FileSystem.copyFileAsync`, which will read the source file from the filesystem for every destination.